tempesta_fw/main.c

```diff
@@ -19,103 +20,52 @@
  */
 #include <linux/kernel.h>
 #include <linux/module.h>
-#include <linux/moduleparam.h>
 
 #include "tempesta.h"
-#include "cache.h"
-#include "client.h"
-#include "debugfs.h"
-#include "filter.h"
-#include "http.h"
 #include "log.h"
-#include "server.h"
 
 MODULE_AUTHOR(TFW_AUTHOR);
 MODULE_DESCRIPTION("Tempesta FW");
 MODULE_VERSION("0.3.0");
 MODULE_LICENSE("GPL");
 
-TfwCfg tfw_cfg;
+#define INIT_TFW_MOD(mod_name) 			\
+do {						\
+	extern TfwCfgMod mod_name; 		\
+	r = tfw_cfg_mod_register(&mod_name); 	\
+	if (r)					\
+		goto err;			\
+} while (0)
 
-static unsigned int cache_size = 256 * 1024;
-module_param(cache_size, uint, 0444);
-MODULE_PARM_DESC(cache_size, "Maximum cache size in pages");
-
-static char *cache_path = "/opt/tempesta/cache";
-module_param(cache_path, charp, 0444);
-MODULE_PARM_DESC(cache_path, "Path to cache directory");
-
-int tfw_connection_init(void);
-void tfw_connection_exit(void);
-
-int tfw_sched_dummy_init(void);
-void tfw_sched_dummy_exit(void);
+extern int tfw_cfg_mod_if_init(void);
+extern void tfw_cfg_mod_if_exit(void);
 
 static int __init
 tfw_init(void)
 {
 	int r;
 
-	TFW_LOG("Start Tempesta\n");
-
-	/* Initialize tfw_cfg. */
-	init_rwsem(&tfw_cfg.mtx);
-	tfw_cfg.c_size = cache_size;
-	memcpy(tfw_cfg.c_path, cache_path, DEF_PROC_STR_LEN);
+	TFW_LOG("Initializing Tempesta kernel module\n");
 
-	r = tfw_if_init();
-	if (r)
+	r = tfw_cfg_mod_if_init();
+	if (r) {
+		TFW_ERR("can't initialize Tempesta configuration interface\n");
 		return r;
-
-	r = tfw_debugfs_init();
-	if (r)
-		goto err_debugfs;
-
-	r = tfw_cache_init();
-	if (r)
-		goto err_cache;
-
-	r = tfw_http_init();
-	if (r)
-		goto err_http;
-
-#if 0
-	r = tfw_filter_init();
-#endif
-	if (r)
-		goto err_filter;
-
-	r = tfw_server_init();
-	if (r)
-		goto err_server;
-
-	r = tfw_client_init();
-	if (r)
-		goto err_client;
-
-	r = tfw_connection_init();
-	if (r)
-		goto err_connection;
-
+	}
+
+	INIT_TFW_MOD(tfw_mod_cache);
+	INIT_TFW_MOD(tfw_mod_http);
+	INIT_TFW_MOD(tfw_mod_server);
+	INIT_TFW_MOD(tfw_mod_client);
+	INIT_TFW_MOD(tfw_mod_session);
+	INIT_TFW_MOD(tfw_mod_connection);
```

Comment:
It seems connection doesn't have it's own configuration, so it should not be treated as TfwCfgMod. Just call it's tfw_connection_init() function as normal. Please check all other calls. Additional unresonable complexity is always bad.